### PR TITLE
suppliers dependency

### PR DESCRIPTION
**Pull request by lotoussa**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | no
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #165 
| ❓ Documentation  | no

## Description

Each supplier must be totally independent and do not call other provider methods.

For caching purpose we may want to wrap the aws client in our own struct.

Merge of every PR of the ticket.

### CODE REVIEW

**Review by wbeuil — APPROVED** [review]
LGTM, do not merge it for now.

File: pkg/remote/aws/iam_policy_supplier_test.go [hunk 1]
```diff
@@ -18,10 +18,11 @@ import (
 
 	"github.com/aws/aws-sdk-go/service/iam"
 
-	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 
+	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
```
Comment (wbeuil) [inline pkg/remote/aws/iam_policy_supplier_test.go:24]: Should be `mocks`

File: pkg/remote/aws/iam_role_policy_attachment_supplier_test.go [hunk 2]
```diff
@@ -4,25 +4,24 @@ import (
...
 
 	"github.com/aws/aws-sdk-go/service/iam"
 
-	"github.com/cloudskiff/driftctl/test/goldenfile"
-	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 
+	"github.com/cloudskiff/driftctl/test/goldenfile"
+	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
```
Comment (wbeuil) [inline pkg/remote/aws/iam_role_policy_attachment_supplier_test.go:23]: `mocks`

File: pkg/remote/aws/iam_role_supplier_test.go [hunk 3]
```diff
@@ -18,10 +18,11 @@ import (
 
 	"github.com/aws/aws-sdk-go/service/iam"
 
-	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 
+	mocks2 "github.com/cloudskiff/driftctl/test/mocks"
```
Comment (wbeuil) [inline pkg/remote/aws/iam_role_supplier_test.go:24]: `mocks`

File: pkg/remote/aws/internet_gateway_supplier.go [hunk 4]
```diff
@@ -15,21 +15,21 @@ import (
 type InternetGatewaySupplier struct {
 	reader       terraform.ResourceReader
 	deserializer *resource.Deserializer
-	client       ec2iface.EC2API
+	client       repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/internet_gateway_supplier.go:18]: Let's rewrite each `client` to `repo`.

File: pkg/remote/aws/nat_gateway_supplier.go [hunk 5]
```diff
@@ -15,22 +15,21 @@ import (
 type NatGatewaySupplier struct {
 	reader       terraform.ResourceReader
 	deserializer *resource.Deserializer
-	client       ec2iface.EC2API
+	client       repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/nat_gateway_supplier.go:18]: `repo`

File: pkg/remote/aws/route_supplier.go [hunk 6]
```diff
@@ -15,24 +16,23 @@ import (
 type RouteSupplier struct {
 	reader       terraform.ResourceReader
 	deserializer *resource.Deserializer
-	client       ec2iface.EC2API
+	client       repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/route_supplier.go:19]: `repo`

File: pkg/remote/aws/route_table_association_supplier.go [hunk 7]
```diff
@@ -15,24 +16,23 @@ import (
 type RouteTableAssociationSupplier struct {
 	reader       terraform.ResourceReader
 	deserializer *resource.Deserializer
-	client       ec2iface.EC2API
+	client       repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/route_table_association_supplier.go:19]: `repo`

File: pkg/remote/aws/route_table_supplier.go [hunk 8]
```diff
@@ -19,26 +19,25 @@ import (
 type RouteTableSupplier struct {
 	reader                  terraform.ResourceReader
 	deserializer            *resource.Deserializer
-	client                  ec2iface.EC2API
+	client                  repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/route_table_supplier.go:22]: `repo`

File: pkg/remote/aws/subnet_supplier.go [hunk 9]
```diff
@@ -18,37 +18,23 @@ import (
 type SubnetSupplier struct {
 	reader              terraform.ResourceReader
 	deserializer        *resource.Deserializer
-	client              ec2iface.EC2API
+	client              repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/subnet_supplier.go:21]: `repo`

File: pkg/remote/aws/vpc_security_group_rule_supplier.go [hunk 10]
```diff
@@ -24,7 +24,7 @@ const (
 type VPCSecurityGroupRuleSupplier struct {
 	reader       terraform.ResourceReader
 	deserializer *resource.Deserializer
-	client       ec2iface.EC2API
+	client       repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/vpc_security_group_rule_supplier.go:27]: `repo`

File: pkg/remote/aws/vpc_security_group_supplier.go [hunk 11]
```diff
@@ -10,31 +11,30 @@ import (
...
-	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
 	"github.com/sirupsen/logrus"
 	"github.com/zclconf/go-cty/cty"
 )
 
 type VPCSecurityGroupSupplier struct {
 	reader                     terraform.ResourceReader
 	deserializer               *resource.Deserializer
-	client                     ec2iface.EC2API
+	client                     repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/vpc_security_group_supplier.go:21]: `repo`

File: pkg/remote/aws/vpc_supplier.go [hunk 12]
```diff
@@ -18,37 +18,23 @@ import (
 type VPCSupplier struct {
 	reader           terraform.ResourceReader
 	deserializer     *resource.Deserializer
-	client           ec2iface.EC2API
+	client           repository.EC2Repository
```
Comment (wbeuil) [inline pkg/remote/aws/vpc_supplier.go:21]: `repo`

File: pkg/remote/aws/init.go [hunk 13]
```diff
@@ -69,7 +69,7 @@ func Init(version string, alerter *alerter.Alerter,
 	supplierLibrary.AddSupplier(NewLambdaFunctionSupplier(provider, deserializer, lambdaRepository))
 	supplierLibrary.AddSupplier(NewDBSubnetGroupSupplier(provider, deserializer, rdsRepository))
 	supplierLibrary.AddSupplier(NewDBInstanceSupplier(provider, deserializer, rdsRepository))
-	supplierLibrary.AddSupplier(NewVPCSecurityGroupSupplier(provider, deserializer))
+	supplierLibrary.AddSupplier(NewVPCSecurityGroupSupplier(provider, deserializer, ec2repository))
 	supplierLibrary.AddSupplier(NewIamUserSupplier(provider, deserializer))
```
Comment (wbeuil) [inline pkg/remote/aws/init.go:73]: Just like what you did for `ec2repo`, can you add `iamrepo` as an argument to each IAM supplier here.

**Review by wbeuil — CHANGES_REQUESTED** [review]

File: pkg/remote/aws/repository/ec2_repository.go [hunk 14]
```diff
@@ -16,6 +16,12 @@ type EC2Repository interface {
 	ListAllAddressesAssociation() ([]string, error)
 	ListAllInstances() ([]*ec2.Instance, error)
 	ListAllKeyPairs() ([]*ec2.KeyPairInfo, error)
+	ListAllInternetGateways() ([]*ec2.InternetGateway, error)
+	ListAllSubnets() ([]*ec2.Subnet, []*ec2.Subnet, error)
+	ListAllNatGateways() ([]*ec2.NatGateway, error)
+	ListAllRouteTables() ([]*ec2.RouteTable, error)
+	ListAllVPCs() ([]*ec2.Vpc, []*ec2.Vpc, error)
+	ListAllSecurityGroups() ([]*ec2.SecurityGroup, []*ec2.SecurityGroup, error)
```
Comment (sundowndev) [inline pkg/remote/aws/repository/ec2_repository.go:24]: Can you add cache for those methods?

**Review by sundowndev — CHANGES_REQUESTED** [review]

**Review by sundowndev — APPROVED** [review]

**Review by wbeuil — APPROVED** [review]

**Pull request merged**